src/main/java/com/artipie/rpm/meta/XmlMaid.java

```diff
@@ -0,0 +1,122 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.rpm.meta;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.xml.namespace.QName;
+import javax.xml.stream.XMLEventReader;
+import javax.xml.stream.XMLEventWriter;
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLOutputFactory;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.events.XMLEvent;
+
+/**
+ * Xml maid.
+ * @since 0.3
+ */
+public interface XmlMaid {
+
+    /**
+     * Cleans xml by ids (checksums) and returns actual package count.
+     * @param ids Checksums
+     * @return Packages count
+     * @throws IOException When smth wrong
+     */
+    long clean(List<String> ids) throws IOException;
+
+    /**
+     * Cleans xml by pkgid attribute in package tag.
+     * @since 0.3
+     */
+    final class ByPkgidAttr implements XmlMaid {
+
+        /**
+         * Package tag name.
+         */
+        private static final String TAG = "package";
+
+        /**
+         * Source, what to clear.
+         */
+        private final Path source;
+
+        /**
+         * Where to write clean data.
+         */
+        private final Path target;
+
+        /**
+         * Ctor.
+         * @param source What to clear
+         * @param target Where to write
+         */
+        public ByPkgidAttr(final Path source, final Path target) {
+            this.source = source;
+            this.target = target;
+        }
+
+        @Override
+        public long clean(final List<String> ids) throws IOException {
+            try (InputStream in = Files.newInputStream(this.source);
+                OutputStream out = Files.newOutputStream(this.target)) {
+                final XMLEventReader reader =
+                    XMLInputFactory.newInstance().createXMLEventReader(in);
+                final XMLEventWriter writer =
+                    XMLOutputFactory.newInstance().createXMLEventWriter(out);
+                XMLEvent event;
+                boolean valid = true;
+                final AtomicLong cnt = new AtomicLong();
+                while (reader.hasNext()) {
+                    event = reader.nextEvent();
+                    if (event.isStartElement()
+                        && event.asStartElement().getName().getLocalPart().equals(ByPkgidAttr.TAG)
+                    ) {
+                        cnt.incrementAndGet();
+                        valid = !ids.contains(
```

Comment:
@olenagerasimova In general, an approach using xslt, or at least java with xpath queries would be more readable. Anyway could we extract event.asStartElement() to a variable? or also may be `element.getName().getLocalPart().equals(ByPkgidAttr.TAG)` to a method (isPackage ?)

Reply:
@HDouss there is no point in the method you are suggesting unfortunately, because I will not be able to use it twice: `StartElement` and `EndElement` both extends `XMLEvent` only and `XMLEvent` does not have `getName()` method. As for the variable, I personally hate not inlined variable...
I've extracted whole `while` intro separate method.

Reply:
@olenagerasimova Got it, personally I prefer to extract to a variable every expression used more than once.